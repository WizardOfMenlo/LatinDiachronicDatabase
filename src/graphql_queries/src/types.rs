use crate::context::Context;
use chrono::prelude::Datelike;
use query_system::ids::AuthorId;
use query_system::ids::FormDataId;
use query_system::ids::FormId;
use query_system::ids::LemmaId;
use query_system::lit_subset::LitSubset;
use query_system::traits::*;
use query_system::types;

pub struct Author {
    id: AuthorId,
}

pub struct TimeSpan {
    time_span: authors_chrono::TimeSpan,
}

#[juniper::object]
impl TimeSpan {
    fn start(&self) -> i32 {
        self.time_span.start().year()
    }

    fn end(&self) -> i32 {
        self.time_span.end().year()
    }
}

impl Author {
    pub(crate) fn new(id: AuthorId) -> Self {
        Author { id }
    }

    fn author(&self, context: &Context) -> types::Author {
        context
            .get()
            .authors()
            .get_by_right(&self.id)
            .cloned()
            .expect("No authorid should be created")
    }
}

#[juniper::object(Context = Context)]
impl Author {
    fn name(&self, context: &Context) -> String {
        self.author(context).name().to_string()
    }

    fn sources(&self, context: &Context) -> Vec<Source> {
        let db = context.get();
        let sources = db.associated_sources(self.id);
        db.sources()
            .iter()
            .filter(|(_, v)| sources.contains(v))
            .map(|(k, _)| Source::new(k.as_path()))
            .collect()
    }

    fn time_span(&self, context: &Context) -> Option<TimeSpan> {
        let auth = self.author(context);
        auth.tspan()
            .cloned()
            .map(|time_span| TimeSpan { time_span })
    }
}

#[derive(juniper::GraphQLObject)]
pub struct Source {
    name: String,
}

impl Source {
    fn new(p: &std::path::Path) -> Self {
        Source {
            name: p.file_name().unwrap().to_string_lossy().to_string(),
        }
    }
}

pub struct Occurrence {
    id: FormDataId,
}

#[juniper::object(Context = Context)]
impl Occurrence {
    fn line(&self, context: &Context) -> String {
        let db = context.get();
        let fd = db.lookup_intern_form_data(self.id);
        db.get_line(fd.source(), fd.line_no()).unwrap().to_string()
    }

    fn source(&self, context: &Context) -> Source {
        let db = context.get();
        let fd = db.lookup_intern_form_data(self.id);

        // TODO, this code is duplicated, extract and refactor
        db.sources()
            .iter()
            .find(|(_, &v)| v == fd.source())
            .map(|(k, _)| Source::new(k))
            .unwrap()
    }
}

pub struct Form {
    form: FormId,
    authors: Vec<AuthorId>,
}

impl Form {
    pub(crate) fn new(form: FormId, authors: impl IntoIterator<Item = AuthorId>) -> Self {
        Form {
            form,
            authors: authors.into_iter().collect(),
        }
    }
}

#[juniper::object(Context = Context)]
impl Form {
    fn form(&self, context: &Context) -> String {
        context
            .get()
            .lookup_intern_form(self.form)
            .0
            .inner()
            .to_string()
    }

    fn count(&self, context: &Context) -> i32 {
        let db = context.get();
        db.count_form_occurrences_subset(
            self.form,
            LitSubset::from_authors(self.authors.iter(), &db),
        ) as i32
    }

    fn occurrences(&self, context: &Context) -> Vec<Occurrence> {
        let db = context.get();
        db.form_occurrences_subset(self.form, LitSubset::from_authors(self.authors.iter(), &db))
            .iter()
            .map(|s| Occurrence { id: *s })
            .collect()
    }
}

pub struct Lemma {
    lemma: LemmaId,
    authors: Vec<AuthorId>,
}

impl Lemma {
    pub(crate) fn new(lemma: LemmaId, authors: impl IntoIterator<Item = AuthorId>) -> Self {
        Lemma {
            lemma,
            authors: authors.into_iter().collect(),
        }
    }
}

#[juniper::object(Context = Context)]
impl Lemma {
    fn lemma(&self, context: &Context) -> String {
        context
            .get()
            .lookup_intern_lemma(self.lemma)
            .0
            .inner()
            .to_string()
    }

    fn count(&self, context: &Context) -> i32 {
        let db = context.get();
        db.count_lemma_occurrences_subset(
            self.lemma,
            LitSubset::from_authors(self.authors.iter(), &db),
        ) as i32
    }

    fn occurrences(&self, context: &Context) -> Vec<Occurrence> {
        let db = context.get();
        db.lemma_occurrences_subset(
            self.lemma,
            LitSubset::from_authors(self.authors.iter(), &db),
        )
        .iter()
        .map(|s| Occurrence { id: *s })
        .collect()
    }
}
