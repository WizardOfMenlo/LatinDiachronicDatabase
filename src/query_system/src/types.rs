//! This mod defines all the structs used for computation.
//! In most cases, the types are created by direct computation on
//! the sources, and are then interned in order to speed up computation

use crate::ids::{AuthorId, FormDataId, FormId, LemmaId, SourceId};

use crate::traits::MainDatabase;
use latin_utilities::NormalizedLatinString;
#[salsa::query_group(InternersGroup)]
pub trait InternDatabase {
    #[salsa::interned]
    fn intern_form_data(&self, fd: FormData) -> FormDataId;

    #[salsa::interned]
    fn intern_form(&self, fd: Form) -> FormId;

    #[salsa::interned]
    fn intern_lemma(&self, fd: Lemma) -> LemmaId;
}

pub trait AuthorInternDatabase {
    fn intern_author(&mut self, author: Author) -> AuthorId;
    fn lookup_intern_author(&self, id: AuthorId) -> &Author;
}

#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct FormData {
    source: SourceId,
    line_no: usize,
    form: FormId,
}

#[derive(shrinkwraprs::Shrinkwrap, Debug, Hash, Eq, PartialEq, Clone, Ord, PartialOrd)]
pub struct Lemma(pub NormalizedLatinString);

#[derive(shrinkwraprs::Shrinkwrap, Debug, Hash, Eq, PartialEq, Clone, Ord, PartialOrd)]
pub struct Form(pub NormalizedLatinString);

pub use authors_chrono::Author;

impl FormData {
    pub fn new(source: SourceId, line_no: usize, form: FormId) -> Self {
        Self {
            source,
            line_no,
            form,
        }
    }

    pub fn source(&self) -> SourceId {
        self.source
    }

    pub fn line_no(&self) -> usize {
        self.line_no
    }

    pub fn form(&self) -> FormId {
        self.form
    }

    pub fn author(&self, db: &impl MainDatabase) -> AuthorId {
        db.associated_author(self.source())
    }
}
