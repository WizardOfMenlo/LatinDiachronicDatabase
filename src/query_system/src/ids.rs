//! The mod where all the `ids` live
//! This encompasses both the interned ids and the ones used throught, such as `AuthorId` and `SourceId`

use salsa::InternId;
use salsa::InternKey;

// Convenience, since all of these are repeated almost exactly
macro_rules! create_ids {
    ($($name:ident),*) => {
        $(
        #[derive(Debug, Hash, Eq, Copy, PartialEq, Clone, PartialOrd, Ord)]
        pub struct $name(InternId);
        impl_intern_key!($name);
        )*
    };
}

macro_rules! impl_intern_key {
    ($name:ident) => {
        impl salsa::InternKey for $name {
            fn from_intern_id(v: InternId) -> Self {
                $name(v)
            }

            fn as_intern_id(&self) -> InternId {
                self.0
            }
        }

        impl $name {
            pub fn from_integer(v: u32) -> Self {
                $name::from_intern_id(InternId::from(v))
            }
        }
    };
}

create_ids!(AuthorId, FormDataId, FormId, LemmaId, SourceId);
