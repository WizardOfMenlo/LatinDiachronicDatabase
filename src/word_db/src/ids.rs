use interner::{RawId, impl_arena_id};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FormId(RawId);
impl_arena_id!(FormId);

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct LemmaId(RawId);
impl_arena_id!(LemmaId);

