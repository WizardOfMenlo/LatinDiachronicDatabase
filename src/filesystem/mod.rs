use super::query_system::ids::SourceId;

use bimap::BiMap;
//use notify::{EventFn, RecommendedWatcher, RecursiveMode, Watcher};
use std::path::PathBuf;

pub trait FileSystem {
    type Source;

    fn load(&self, id: SourceId) -> String;
    fn watch(&self, id: SourceId);

    fn intern_source(&mut self, path: Self::Source) -> SourceId;
}

// TODO: This might just be AsRef
pub trait GetFileSystem {
    type Fs: FileSystem;

    fn filesystem(&self) -> &Self::Fs;
    fn filesystem_mut(&mut self) -> &mut Self::Fs;
}

impl<T> FileSystem for T
where
    T: GetFileSystem,
{
    type Source = <T::Fs as FileSystem>::Source;

    fn intern_source(&mut self, path: Self::Source) -> SourceId {
        self.filesystem_mut().intern_source(path)
    }

    fn watch(&self, id: SourceId) {
        self.filesystem().watch(id)
    }

    fn load(&self, id: SourceId) -> String {
        self.filesystem().load(id)
    }
}

#[derive(Debug, Default, Clone)]
pub struct InternerFileSystem {
    sources: BiMap<PathBuf, SourceId>,
}

impl InternerFileSystem {
    pub fn new() -> Self {
        InternerFileSystem {
            sources: BiMap::new(),
        }
    }

    pub fn sources(&self) -> &BiMap<PathBuf, SourceId> {
        &self.sources
    }
}

impl FileSystem for InternerFileSystem {
    type Source = PathBuf;

    fn intern_source(&mut self, path: Self::Source) -> SourceId {
        if self.sources.contains_left(&path) {
            return *self.sources.get_by_left(&path).unwrap();
        }

        // Guarantee that we never overwrite a value
        let mut new_id = self.sources.len() as u32;
        while self.sources.contains_right(&SourceId::from_integer(new_id)) {
            new_id += 1;
        }

        let new_id = SourceId::from_integer(new_id);
        self.sources.insert(path, new_id);
        new_id
    }

    fn watch(&self, id: SourceId) {
        let _path = self.sources.get_by_right(&id).unwrap();
        /*
        // TODO: Error recovery
        self.watcher
            .watch(path, RecursiveMode::NonRecursive)
            .unwrap();
        */
    }

    fn load(&self, id: SourceId) -> String {
        // TODO: Check conversions
        std::fs::read_to_string(self.sources.get_by_right(&id).unwrap()).unwrap()
    }
}

#[derive(Debug, Default, Clone)]
pub struct MockFileSystem {
    strings: BiMap<String, SourceId>,
}

impl FileSystem for MockFileSystem {
    type Source = String;

    fn intern_source(&mut self, path: Self::Source) -> SourceId {
        // TODO: Bound check
        let new_id = SourceId::from_integer(self.strings.len() as u32);
        self.strings.insert(path, new_id);
        new_id
    }

    fn watch(&self, _: SourceId) {}

    fn load(&self, id: SourceId) -> String {
        self.strings.get_by_right(&id).unwrap().clone()
    }
}
