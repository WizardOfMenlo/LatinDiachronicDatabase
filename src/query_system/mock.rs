//! Mocking facilities for testing

use super::gc::GCollectable;
use super::ids::AuthorId;
use super::middle::IntermediateQueries;
use super::sources::SourcesQueryGroup;
use super::traits::{AuthorInternDatabase, IntermediateDatabase};
use super::types::InternersGroup;
use super::MainQueries;
use crate::authors_chrono::Author;
use crate::latin_lemmatizer::NaiveLemmatizer;

use std::collections::HashMap;
use std::sync::Arc;

/// A simplified database, which we use for testing
#[salsa::database(SourcesQueryGroup, InternersGroup, MainQueries, IntermediateQueries)]
pub struct MockDatabase {
    runtime: salsa::Runtime<MockDatabase>,
    mock: Author,
}

/// A mock database to be used for preliminary testing
/// Note, we have empty lemm, so this will fail complex queries
pub fn make_mock() -> MockDatabase {
    let mut res = MockDatabase::new();
    res.set_lemmatizer(Arc::new(NaiveLemmatizer::new(HashMap::new())));
    res
}

impl MockDatabase {
    pub fn new() -> Self {
        MockDatabase {
            runtime: salsa::Runtime::default(),
            mock: Author::new("Mock"),
        }
    }
}

impl AuthorInternDatabase for MockDatabase {
    fn intern_author(&mut self, _: Author) -> AuthorId {
        AuthorId::from_integer(0)
    }

    fn lookup_intern_author(&self, _: AuthorId) -> &Author {
        &self.mock
    }
}

impl Default for MockDatabase {
    fn default() -> Self {
        Self::new()
    }
}

impl GCollectable for MockDatabase {
    fn garbage_sweep(&mut self) {}
}

impl salsa::Database for MockDatabase {
    fn salsa_runtime(&self) -> &salsa::Runtime<Self> {
        &self.runtime
    }

    fn salsa_runtime_mut(&mut self) -> &mut salsa::Runtime<Self> {
        &mut self.runtime
    }
}

impl salsa::ParallelDatabase for MockDatabase {
    fn snapshot(&self) -> salsa::Snapshot<Self> {
        salsa::Snapshot::new(MockDatabase {
            runtime: self.runtime.snapshot(self),
            mock: self.mock.clone(),
        })
    }
}
