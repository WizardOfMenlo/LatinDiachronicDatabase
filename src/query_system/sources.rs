//! The lowest level interfacing with source files directly

use super::ids::{AuthorId, FormDataId, SourceId};
use super::types::{Form, FormData, InternDatabase};
use crate::filesystem::FileSystem;
use crate::latin_utilities::StandardLatinConverter;

use log::info;
use std::collections::HashSet;
use std::sync::Arc;

/// The trait that is used to parse sources
/// Usage: set the source text, define the relation between sources and authors
#[salsa::query_group(SourcesQueryGroup)]
pub trait SourcesDatabase: InternDatabase + FileSystem + salsa::Database {
    /// Get the source text for a specified source
    fn source_text(&self, source_id: SourceId) -> Arc<String>;

    /// Get the sources for an author
    #[salsa::input]
    fn associated_sources(&self, author_id: AuthorId) -> Arc<HashSet<SourceId>>;

    /// Get the author for a source
    #[salsa::input]
    fn associated_author(&self, source_id: SourceId) -> AuthorId;

    // Low level
    /// Get a determined line in a source, if possible
    fn get_line(&self, source_id: SourceId, line: usize) -> Option<Arc<String>>;

    // TODO, benchmark and see if hashset actually worth it
    /// Parse a source, returning the FormData that it generates
    fn parse_source(&self, source_id: SourceId) -> Arc<HashSet<FormDataId>>;
}

fn source_text(db: &impl SourcesDatabase, source_id: SourceId) -> Arc<String> {
    info!("Loading source {:?}", source_id);
    db.salsa_runtime()
        .report_synthetic_read(salsa::Durability::LOW);
    db.watch(source_id);
    Arc::new(db.load(source_id))
}

// Note, this function is O(line), so it should be used scarcely
fn get_line(db: &impl SourcesDatabase, source_id: SourceId, line: usize) -> Option<Arc<String>> {
    let text = db.source_text(source_id);
    text.lines().nth(line).map(|l| Arc::new(l.to_string()))
}

fn parse_source(db: &impl SourcesDatabase, source_id: SourceId) -> Arc<HashSet<FormDataId>> {
    info!("Parsing source {:?}", source_id);
    let converter = StandardLatinConverter::default();
    let mut form_data_ids = HashSet::new();

    let text = db.source_text(source_id);

    for (i, line) in text.lines().enumerate() {
        for word in line.split(' ') {
            let lw = converter.convert(word);
            let form = Form(lw);
            let form_id = db.intern_form(form);
            let form_data = FormData::new(source_id, i, form_id);
            let form_data_id = db.intern_form_data(form_data);
            form_data_ids.insert(form_data_id);
        }
    }

    Arc::new(form_data_ids)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::filesystem::FileSystem;
    use crate::query_system::mock::make_mock;
    use proptest::prelude::*;
    use std::iter;

    fn generate_source_n_lines<T: ToString>(gen: impl Fn(usize) -> T, n: usize) -> String {
        iter::repeat(())
            .enumerate()
            .map(|(i, ())| gen(i).to_string())
            .take(n)
            .collect::<Vec<_>>()
            .join("\n")
    }

    #[test]
    #[should_panic]
    fn test_panic_non_set_source_text() {
        let db = make_mock();
        db.source_text(SourceId::from_integer(0));
    }

    #[test]
    #[should_panic]
    fn test_panic_non_set_author_sources() {
        let db = make_mock();
        db.associated_sources(AuthorId::from_integer(0));
    }

    #[test]
    fn test_empty_source_parsing() {
        let mut db = make_mock();
        let source = db.intern_source(String::new());
        let res = db.parse_source(source);

        assert_eq!(res.len(), 0);
    }

    #[test]
    fn test_source_parsing() {
        let mut db = make_mock();
        let source = db.intern_source(generate_source_n_lines(|_| "puella", 100));
        let parse_res = db.parse_source(source);

        assert_eq!(parse_res.len(), 100);

        let form_data: Vec<_> = parse_res
            .iter()
            .map(|&fd| db.lookup_intern_form_data(fd))
            .collect();
        let form_id = form_data[0].form();

        assert!(form_data.iter().all(|fd| fd.source() == source));
        assert!(form_data.iter().all(|fd| fd.form() == form_id));
    }

    #[test]
    fn test_get_line() {
        let mut db = make_mock();
        let source = db.intern_source(generate_source_n_lines(|i| i, 100));
        for i in 0..100 {
            let line = db.get_line(source, i).expect("Line should have been set");
            // Note, since the line is stored as string, not as a normalized one, this should always succeed.
            assert_eq!(*line, i.to_string());
        }
    }

    use insta::assert_debug_snapshot;
    use std::collections::BTreeSet;

    #[test]
    fn parse_lorem_ipsum() {
        let mut db = make_mock();
        let source = db.intern_source(String::from(
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."));
        let parse_res = db.parse_source(source);
        let form_data: BTreeSet<_> = parse_res
            .iter()
            .map(|&fd| db.lookup_intern_form_data(fd))
            .map(|fd| (db.lookup_intern_form(fd.form()), fd.source(), fd.line_no()))
            .collect();

        assert_debug_snapshot!("lorem_ipusm", form_data)
    }

    #[test]
    fn parse_segment() {
        let text = r#"Notum est omnibus, Celse, penes te studiorum
nostrorum manere summam, ideoque primum sedulitatis meae
inpendium iudiciis tuis offerre proposui. nam cum sibi
inter aequales quendam locum deposcat aemulatio,
neminem magis conatibus nostris profuturum credidi quam
qui inter eos in hac parte plurimum possit. itaque quo
cultior in quorundam notitiam ueniat, omnia tibi nota
perlaturus ad te primum liber iste festinet, apud te"#;

        let mut db = make_mock();
        let source = db.intern_source(text.to_string());
        let parse_res = db.parse_source(source);
        let form_data: BTreeSet<_> = parse_res
            .iter()
            .map(|&fd| db.lookup_intern_form_data(fd))
            .map(|fd| (db.lookup_intern_form(fd.form()), fd.source(), fd.line_no()))
            .collect();

        assert_debug_snapshot!("notum_est_omnibus", form_data)
    }

    proptest! {
        #[test]
        fn doesnt_crash(s in "\\PC*") {
            let mut db = make_mock();
            let source = db.intern_source(s);
            let _parse_res = db.parse_source(source);
        }
    }
}
