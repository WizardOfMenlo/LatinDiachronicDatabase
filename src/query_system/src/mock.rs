//! Mocking facilities for testing

use crate::ids::AuthorId;
use crate::middle::IntermediateQueries;
use crate::sources::SourcesQueryGroup;
use crate::traits::{AuthorInternDatabase, IntermediateDatabase};
use crate::types::InternersGroup;
use crate::MainQueries;
use authors_chrono::Author;
use latin_lemmatizer::NaiveLemmatizer;
use std::collections::HashMap;
use std::sync::Arc;

/// A simplified database, which we use for testing
#[salsa::database(SourcesQueryGroup, InternersGroup, MainQueries, IntermediateQueries)]
pub struct MockDatabase {
    runtime: salsa::Runtime<MockDatabase>,
    mock: Author,
}

/// A mock database to be used for preliminary testing
/// Note, we have empty lemm, so this will fail complex queries
pub fn make_mock() -> MockDatabase {
    let mut res = MockDatabase::new();
    res.set_lemmatizer(Arc::new(NaiveLemmatizer::new(HashMap::new())));
    res
}

impl MockDatabase {
    pub fn new() -> Self {
        MockDatabase {
            runtime: salsa::Runtime::default(),
            mock: Author::new("Mock"),
        }
    }
}

impl AuthorInternDatabase for MockDatabase {
    fn intern_author(&mut self, _: Author) -> AuthorId {
        AuthorId::from_integer(0)
    }

    fn lookup_intern_author(&self, _: AuthorId) -> &Author {
        &self.mock
    }
}

impl Default for MockDatabase {
    fn default() -> Self {
        Self::new()
    }
}

impl salsa::Database for MockDatabase {
    fn salsa_runtime(&self) -> &salsa::Runtime<Self> {
        &self.runtime
    }
}

impl salsa::ParallelDatabase for MockDatabase {
    fn snapshot(&self) -> salsa::Snapshot<Self> {
        salsa::Snapshot::new(MockDatabase {
            runtime: self.runtime.snapshot(self),
            mock: self.mock.clone(),
        })
    }
}
