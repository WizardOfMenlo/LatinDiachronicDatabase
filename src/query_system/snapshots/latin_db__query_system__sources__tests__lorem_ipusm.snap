---
source: src/query_system/sources.rs
expression: form_data
---
{
    (
        Form(
            NormalizedLatinString(
                "ad",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "adipiscing",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "aliqua",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "aliquip",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "amet",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "anim",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "aute",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "cillum",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "commodo",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "consectetur",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "consequat",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "culpa",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "cupidatat",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "deserunt",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "do",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "dolor",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "dolore",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "duis",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ea",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "eiusmod",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "elit",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "enim",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "esse",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "est",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "et",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "eu",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ex",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "excepteur",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "exercitation",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "fugiat",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "id",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "in",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "incididunt",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ipsum",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "irure",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "labore",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "laboris",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "laborum",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "lorem",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "magna",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "minim",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "mollit",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "nisi",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "non",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "nostrud",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "nulla",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "occaecat",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "officia",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "pariatur",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "proident",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "qui",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "quis",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "reprehenderit",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "sed",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "sint",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "sit",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "sunt",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "tempor",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "uelit",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ueniam",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ullamco",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "uoluptate",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
    (
        Form(
            NormalizedLatinString(
                "ut",
            ),
        ),
        SourceId(
            0,
        ),
        0,
    ),
}
